fhir-persistence-cos/src/main/java/com/ibm/fhir/persistence/cos/impl/COSClientManager.java

```diff
@@ -0,0 +1,129 @@
+            throw new IllegalStateException("COSClientManager is shut down");
+        }
+
+        // Connections can be tenant-specific, so find out what tenant we're associated with and use its persistence
+        // configuration to obtain the appropriate CqlSession instance (shared by multiple threads).
+        final String tenantId = FHIRRequestContext.get().getTenantId();
+        final String dsId = "default"; // only one payload datasource
+        TenantDatasourceKey key = new TenantDatasourceKey(tenantId, dsId);
+
+        // Get the session for this tenant/datasource, or create a new one if needed
+        return clientMap.computeIfAbsent(key, COSClientManager::newClient);
+    }
+
+    /**
+     * Build a new CqlSession object for the tenant/datasource.
+     * @param key
+     * @return
+     */
+    private static COSPayloadClient newClient(TenantDatasourceKey key) {
+
+        String dsPropertyName = FHIRConfiguration.PROPERTY_PERSISTENCE_PAYLOAD + "/" + key.getDatasourceId();
+        PropertyGroup dsPG = FHIRConfigHelper.getPropertyGroup(dsPropertyName);
+        if (dsPG == null) {
+            throw new IllegalStateException("Could not locate configuration property group: " + dsPropertyName);
+        }
+
+        try {
+            // Get the payload datasource type (cassandra/cos).
+            String type = dsPG.getStringProperty("type", null);
+            if (type == null) {
+                throw new IllegalStateException("Could not locate 'type' property within datasource property group: " + dsPropertyName);
+            }
+
+            // Confirm that this is a COS datasource configuration element
+            if (!"cos".equals(type)) {
+                throw new IllegalStateException("Unsupported 'type' property value within datasource property group: " + type);
+            }
+
+            // Get the connection properties
+            PropertyGroup connectionProps = dsPG.getPropertyGroup("connectionProperties");
+            if (connectionProps == null) {
+                throw new IllegalStateException("Could not locate 'connectionProperties' property group within datasource property group: " + dsPropertyName);
+            }
+
+            // Wrap the connection properties in an adapter to simplify access
+            CosPropertyGroupAdapter adapter = new CosPropertyGroupAdapter(connectionProps);
+            return new COSPayloadClient(key.getTenantId(), adapter);
+        } catch (Exception x) {
+            throw new IllegalStateException(x);
+        }
+    }
+
+    /**
+     * Get rid of any sessions we're holding on to
+     */
+    private void closeAllClients() {
+        // prevent anyone asking for a client
+        this.running = false;
+        clientMap.clear();
+    }
+
+    /**
+     *
+     */
+    public static void shutdown() {
+        logger.info("Shutting down COSClientManager");
+        getInstance().closeAllClients();
+        logger.info("COSClientManager shutdown complete");
+    }
```

Comment:
once again, does this need to hook tot he ServletContextListener? is it only intended to be commandline?